### PR TITLE
Correct lifetime based cancellation of AsyncStream and AsyncThrowingStream

### PR DESCRIPTION
**Pull request by phausler**
This addresses an outstanding issue where AsyncStream and AsyncThrowingStream could not terminate in the case of indefinite sequences for a handler that lived by capture. AsyncStream and AsyncThrowingStream now hold a context that cancels on deinitialization. 

Resolves radar: rdar://82985344

Tests incoming soon.

### CODE REVIEW

**Comment by phausler** [comment]
@swift-ci please smoke test

**Comment by phausler** [comment]
@swift-ci please smoke test

File: stdlib/public/Concurrency/AsyncThrowingStream.swift [hunk 1]
```diff
@@ -132,14 +146,26 @@ public struct AsyncThrowingStream<Element, Failure: Error> {
...
+      = .create(produce)
+    context = _Context {
+      return try await withTaskCancellationHandler {
+        guard let result = try await storage.value?() else {
+          storage.value = nil
+          return nil
+        }
+        return result
+      } onCancel: {
+        storage.value = nil
```
Comment (ktoso) [inline stdlib/public/Concurrency/AsyncThrowingStream.swift:166]: okey and that write is lock protected... looks good

**Review by ktoso — APPROVED** [review]
LGTM

**Pull request merged**